ibis/common/grounds.py

```diff
@@ -117,78 +106,74 @@ def __new__(metacls, clsname, bases, dct):
         return super().__new__(metacls, clsname, bases, attribs)
 
 
-class Annotable(Base, Immutable, metaclass=AnnotableMeta):
+class Annotable(Base, metaclass=AnnotableMeta):
     """Base class for objects with custom validation rules."""
 
-    __slots__ = ("__args__", "__precomputed_hash__")
-
     @classmethod
     def __create__(cls, *args, **kwargs):
         # construct the instance by passing the validated keyword arguments
         kwargs = cls.__signature__.validate(*args, **kwargs)
         return super().__create__(**kwargs)
 
     def __init__(self, **kwargs):
-        # set the already validated fields using object.__setattr__ since we
-        # treat the annotable instances as immutable objects
+        # set the already validated fields using object.__setattr__
         for name, value in kwargs.items():
             object.__setattr__(self, name, value)
+        # allow child classes to do some post-initialization
+        self.__post_init__()
 
-        # optimizations to store frequently accessed generic properties
-        args = tuple(kwargs[name] for name in self.__argnames__)
-        object.__setattr__(self, "__args__", args)
-        object.__setattr__(
-            self, "__precomputed_hash__", hash((self.__class__, args))
-        )
-
+    def __post_init__(self):
         # calculate special property-like objects only once due to the
         # immutable nature of annotable instances
         for name, prop in self.__properties__.items():
             object.__setattr__(self, name, prop(self))
 
-        # any supplemental custom code provided by descendant classes
-        self.__post_init__()
-
-    def __post_init__(self):
-        pass
+    def __setattr__(self, name, value):
+        param = self.__signature__.parameters[name]
+        if not (param.default is None and value is None):
```

Comment:
```suggestion
        if param.default is not None or value is not None:
```

I personally find it difficult to read `not (complicated boolean expression)` versus its De Morgan equivalent without the initial `not`.

Reply:
Going to update.